tensorflow/c/eager/dlpack.cc

```diff
@@ -0,0 +1,342 @@
+};
+
+void deleter(DLManagedTensor* arg) {
+  TFDLMTensor* owner = static_cast<TFDLMTensor*>(arg->manager_ctx);
+  owner->handle->Unref();
+  delete owner;
+}
+
+DLDataType getDLDataType(TF_DataType data_type, TF_Status* status) {
+  DLDataType dtype;
+  dtype.lanes = 1;
+  dtype.bits = TF_DataTypeSize(data_type) * 8;
+  switch (data_type) {
+    case TF_DataType::TF_FLOAT:
+      dtype.code = DLDataTypeCode::kDLFloat;
+      break;
+    case TF_DataType::TF_DOUBLE:
+      dtype.code = DLDataTypeCode::kDLFloat;
+      break;
+    case TF_DataType::TF_INT32:
+      dtype.code = DLDataTypeCode::kDLInt;
+      break;
+    case TF_DataType::TF_UINT8:
+      dtype.code = DLDataTypeCode::kDLUInt;
+      break;
+    case TF_DataType::TF_INT16:
+      dtype.code = DLDataTypeCode::kDLInt;
+      break;
+    case TF_DataType::TF_STRING:
+      dtype.code = DLDataTypeCode::kDLFloat;
+      break;
+    case TF_DataType::TF_COMPLEX64:
+      status->status = tensorflow::errors::InvalidArgument(
+          "TF_COMPLEX64 is not supported by dlpack");
+      break;
+    case TF_DataType::TF_INT64:
+      dtype.code = DLDataTypeCode::kDLInt;
+      break;
+    case TF_DataType::TF_BOOL:
+      dtype.code = DLDataTypeCode::kDLUInt;
+      break;
+    case TF_DataType::TF_QINT8:
+      status->status = tensorflow::errors::InvalidArgument(
+          "TF_QINT8 is not supported by dlpack");
+      break;
+    case TF_DataType::TF_QUINT8:
+      status->status = tensorflow::errors::InvalidArgument(
+          "TF_QUINT8 is not supported by dlpack");
+      break;
+    case TF_DataType::TF_QINT32:
+      status->status = tensorflow::errors::InvalidArgument(
+          "TF_QINT32 is not supported by dlpack");
+      break;
+    case TF_DataType::TF_BFLOAT16:
+      dtype.code = DLDataTypeCode::kDLBfloat;
+      break;
+    case TF_DataType::TF_QINT16:
+      status->status = tensorflow::errors::InvalidArgument(
+          "TF_QINT16 is not supported by dlpack");
+      break;
+    case TF_DataType::TF_QUINT16:
+      status->status = tensorflow::errors::InvalidArgument(
+          "TF_QUINT16 is not supported by dlpack");
+      break;
+    case TF_DataType::TF_COMPLEX128:
+      status->status = tensorflow::errors::InvalidArgument(
+          "TF_COMPLEX128 is not supported by dlpack");
+      break;
+    case TF_DataType::TF_HALF:
+      dtype.code = DLDataTypeCode::kDLFloat;
+      break;
+    case TF_DataType::TF_RESOURCE:
+      status->status = tensorflow::errors::InvalidArgument(
+          "TF_RESOURCE is not supported by dlpack");
+      break;
+    case TF_DataType::TF_VARIANT:
+      status->status = tensorflow::errors::InvalidArgument(
+          "TF_VARIANT is not supported by dlpack");
+      break;
+    case TF_DataType::TF_UINT32:
+      dtype.code = DLDataTypeCode::kDLUInt;
+      break;
+    case TF_DataType::TF_UINT64:
+      dtype.code = DLDataTypeCode::kDLUInt;
+      break;
+    default:
+      status->status = tensorflow::errors::InvalidArgument(
```

Comment:
You could probably fold a lot of the unsupported cases into default by using DataTypeString to get the dtype name

Reply:
Actually I'm not sure what's the proper way to throw the errors. Could you point me somewhere  I can look at?

Reply:
I meant something like status->status = errors::InvalidArgument(DataTypeString(dtype), " cannot be used with dlpack") or something like that.